### PR TITLE
Unify pattern matching and declaration conditions

### PR DESCRIPTION
**Pull request by STRd6**
- [x] Refactor declaration conditions
- [x] Apply pattern matching conditions to declaration conditions

There's some possible follow on discussion/work to do.

Should

```civet
if {x} := y
  ...
```

also add a condition check for x being non-falsey?

What about:

```
if {x, y} := z
  ...
```

Should x and y be non-falsey?

### CODE REVIEW

File: test/if.civet [hunk 1]
```diff
@@ -282,9 +282,7 @@ describe "if", ->
     ---
     x if { x } := y
     ---
-    let ref;if (ref = y) {
-    const { x } = ref;x
-    }
+    let ref;if ((ref = y) && 'x' in ref) { const { x } = ref;x }
```
Comment (edemaine) [inline test/if.civet:285]: Whoa, this is pretty interesting. But I think makes sense; `:=` is finally becoming a pattern-matching operator.

File: test/if.civet [hunk 2]
```diff
@@ -296,15 +294,13 @@ describe "if", ->
       z
     } if w := g()
     ---
-    let ref;if (ref = g()) {
-    const w = ref;if (ref1 = f()) {
+    let ref;if (ref = g()) { const w = ref;if (ref1 = f()) {
```
Comment (edemaine) [inline test/if.civet:297]: This seems to be missing a `let ref1`.

File: test/if.civet [hunk 3]
```diff
@@ -797,13 +793,25 @@ describe "if", ->
...
+      ---
+      if {x, ^y} := obj
+        console.log x
+      ---
+      let ref;if ((ref = obj) && 'x' in ref && 'y' in ref && ref.y === y) {
+        const {x, y} = ref;
+        console.log(x)
+      }
+    """
+
```
Comment (edemaine) [inline test/if.civet:814]: 🤯 Very cool.

File: source/lib.civet [hunk 4]
```diff
@@ -131,26 +144,18 @@ function updateParentPointers(node: ASTNode, parent?: ASTNodeBase, depth = 1): v
...
-
-  if prefix?.length
-    indent := prefix[0][0]
-    expressions = [...prefix, [indent, statement]]
-    children = [" {\n", ...expressions, "\n", indent?.slice?.(0, -2), "}"]
-  else
-    expressions = [["", statement]]
-    children = [" { ", ...expressions, " }"]
+  expressions := [
+    ...post.blockPrefix or []
```
Comment (edemaine) [inline source/lib.civet:148]: I wonder about the loss of `post.condition?.expression.blockPrefix` here... but maybe it's not the cause of the bug below.

**Review by edemaine — COMMENTED** [review]
This is very cool! I think I found one bug in one of the examples though (my annoying double-`if` example from before).

IMO, `if {x, y} := z` should not check for truthiness of `x` and `y`; I think it should behave the same as
```js
if (
  switch z
    {x, y}
      true
    else
      false
)
```

File: test/if.civet (hunk 2 above)
Comment (STRd6) [reply test/if.civet:297]: It looks like it was missing that from the first time I added it, I'll see about adding it tomorrow.

File: source/lib.civet (hunk 4 above)
Comment (STRd6) [reply source/lib.civet:148]: The condition expression block prefix is now added in a more uniform way inside `processDeclarationConditions`, this one here is to pick up the random prefixes from coffee style for loops until we move them later on in the parse tree processing.

**Pull request merged**